// @nutui/nutui-react-taro/dist/esm/swipe.taro-CUZSZcIh.js
import { a as __awaiter } from "./tslib.es6-iWu3F_1J.js";
import React__default, { forwardRef, useRef, useState, useImperativeHandle, useEffect } from "react";
import classNames from "classnames";
import { View } from "@tarojs/components";
import { useReady, nextTick } from "@tarojs/taro";
import { u as useTouch } from "./use-touch-BeiLHDO9.js";
import { g as getRectByTaro } from "./get-rect-by-taro-D0h6aiDr.js";
import { C as ComponentDefaults } from "./typings-DV9RBfhj.js";
function preventDefault(event, isStopPropagation) {
  if (typeof event.cancelable !== "boolean" || event.cancelable) {
    event.preventDefault();
  }
  {
    event.stopPropagation();
  }
}
const defaultProps = Object.assign(Object.assign({}, ComponentDefaults), { name: "" });
const Swipe = forwardRef((props, instanceRef) => {
  const classPrefix = "nut-swipe";
  const touch = useTouch();
  useReady(() => {
    const getWidth = () => __awaiter(void 0, void 0, void 0, function* () {
      if (leftWrapper.current) {
        const leftRect = yield getRectByTaro(leftWrapper.current);
        leftRect && setActionWidth((v) => Object.assign(Object.assign({}, v), { left: leftRect.width }));
      }
      if (rightWrapper.current) {
        const rightRect = yield getRectByTaro(rightWrapper.current);
        rightRect && setActionWidth((v) => Object.assign(Object.assign({}, v), { right: rightRect.width }));
      }
    });
    nextTick(() => getWidth());
  });
  const { children, className, style } = Object.assign(Object.assign({}, defaultProps), props);
  const root = useRef();
  const opened = useRef(false);
  const lockClick = useRef(false);
  const startOffset = useRef(0);
  const [state, setState] = useState({
    offset: 0,
    dragging: false
  });
  const [actionWidth, setActionWidth] = useState({
    left: 0,
    right: 0
  });
  const wrapperStyle = {
    transform: `translate3d(${state.offset}px, 0, 0)`,
    transitionDuration: state.dragging ? "0s" : ".6s"
  };
  const leftWidth = actionWidth.left;
  const rightWidth = actionWidth.right;
  const onTouchStart = (event) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    if (leftWrapper.current) {
      const leftRect = yield getRectByTaro(leftWrapper.current);
      leftRect && setActionWidth((v) => Object.assign(Object.assign({}, v), { left: leftRect.width }));
    }
    if (rightWrapper.current) {
      const rightRect = yield getRectByTaro(rightWrapper.current);
      rightRect && setActionWidth((v) => Object.assign(Object.assign({}, v), { right: rightRect.width }));
    }
    if (!props.disabled) {
      startOffset.current = state.offset;
      touch.start(event);
      (_a = props.onTouchStart) === null || _a === void 0 ? void 0 : _a.call(props, event);
    }
  });
  const onTouchMove = (event) => {
    var _a;
    if (props.disabled) {
      return;
    }
    touch.move(event);
    (_a = props.onTouchMove) === null || _a === void 0 ? void 0 : _a.call(props, event);
    if (touch.isHorizontal()) {
      lockClick.current = true;
      props.onTouchMove && props.onTouchMove(event);
      const newState = Object.assign(Object.assign({}, state), { dragging: true });
      const isEdge = !opened || touch.deltaX.current * startOffset.current < 0;
      if (isEdge) {
        preventDefault(event);
      }
      newState.offset = rangeCalculation(touch.deltaX.current + startOffset.current, -rightWidth || 0, leftWidth || 0);
      setState(newState);
    }
  };
  const onTouchEnd = (event) => {
    var _a;
    if (state.dragging) {
      setState((v) => Object.assign(Object.assign({}, v), { dragging: false }));
      toggle(state.offset > 0 ? "left" : "right");
      setTimeout(() => {
        lockClick.current = false;
      }, 0);
      (_a = props.onTouchEnd) === null || _a === void 0 ? void 0 : _a.call(props, event);
    }
  };
  const toggle = (side) => {
    const offset = Math.abs(state.offset);
    const base = 0.3;
    const baseNum = opened ? 1 - base : base;
    const width = side === "left" ? leftWidth : rightWidth;
    if (width && offset > Number(width) * baseNum) {
      open(side);
    } else {
      close(side);
    }
  };
  const open = (side) => {
    var _a;
    opened.current = true;
    const offset = side === "left" ? leftWidth : -rightWidth;
    const name = props.name;
    (_a = props.onOpen) === null || _a === void 0 ? void 0 : _a.call(props, { name, position: side });
    setState((v) => Object.assign(Object.assign({}, v), { offset: Number(offset) || 0 }));
  };
  const close = (position) => {
    var _a;
    if (opened.current) {
      opened.current = false;
      (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props, {
        name: props.name,
        position: position || "left"
      });
    }
    setState((v) => Object.assign(Object.assign({}, v), { offset: 0 }));
  };
  const rangeCalculation = (num, min, max) => {
    return Math.min(Math.max(Number(num), Number(min)), Number(max));
  };
  const leftWrapper = useRef(null);
  const rightWrapper = useRef(null);
  const renderActionContent = (side) => {
    if (props[`${side}Action`]) {
      return React__default.createElement("div", { ref: side === "left" ? leftWrapper : rightWrapper, className: `${classPrefix}-${side}`, onClick: (e) => handleOperate(e, side) }, props[`${side}Action`]);
    }
    return null;
  };
  const handleOperate = (event, position) => {
    var _a;
    event.stopPropagation();
    if (props.beforeClose) {
      props.beforeClose(position);
    } else {
      (_a = props.onActionClick) === null || _a === void 0 ? void 0 : _a.call(props, event, position);
    }
  };
  useImperativeHandle(instanceRef, () => ({
    open,
    close: () => close()
  }));
  useEffect(() => {
    const handler = (event) => {
      const targets = [root];
      if (targets.some((targetItem) => {
        const targetElement = targetItem.current || targetItem;
        return !targetElement || (targetElement === null || targetElement === void 0 ? void 0 : targetElement.contains(event.target));
      })) {
        return;
      }
      close();
    };
    document.addEventListener("touchstart", handler);
    return () => {
      document.removeEventListener("touchstart", handler);
    };
  }, []);
  return React__default.createElement(
    View,
    { ref: root, className: classNames(classPrefix, className), onTouchStart: (e) => onTouchStart(e), onTouchMove: (e) => onTouchMove(e), onTouchEnd: (e) => onTouchEnd(e), style },
    React__default.createElement(
      "div",
      { className: `${classPrefix}-wrapper`, style: wrapperStyle },
      renderActionContent("left"),
      children,
      renderActionContent("right")
    )
  );
});
Swipe.displayName = "NutSwipe";
export {
  Swipe as S
};
