// @nutui/nutui-react-taro/dist/esm/video.taro-CecLv4sr.js
import { _ as __rest } from "./tslib.es6-iWu3F_1J.js";
import React__default from "react";
import classNames from "classnames";
import { View, Video as Video$1 } from "@tarojs/components";
const defaultProps = {
  source: {
    type: {},
    src: ""
  },
  options: {
    controls: true,
    muted: false,
    // 默认不是静音
    autoplay: false,
    poster: "",
    playsinline: false,
    loop: false
  }
};
const classPrefix = `nut-video`;
const Video = (props) => {
  const _a = Object.assign(Object.assign({}, defaultProps), props), { children, source, options, className, style, onPlay, onPause, onPlayEnd } = _a, restProps = __rest(_a, ["children", "source", "options", "className", "style", "onPlay", "onPause", "onPlayEnd"]);
  const classes = classNames(classPrefix, className);
  return React__default.createElement(
    View,
    { className: classes, style },
    React__default.createElement(Video$1, Object.assign({ className: "nut-video-player", muted: options.muted, autoplay: options.autoplay, loop: options.loop, poster: options.poster, controls: options.controls, src: source.src, onPlay, onPause, onEnded: onPlayEnd }, restProps))
  );
};
Video.displayName = "NutVideo";
export {
  Video as V
};
